open-apparel-registry/open-apparel-registry
# Upgrade Google Analytics to gtag.js

## Overview

`analytics.js`, an older version of Google Analytics, can't be used with Google Analytics 3. In July 2023, Universal Analytics, which we previously used, will no longer be available for use. In order to enable GA4, we switch to using `gtag.js`.

Connects #2188 

## Demo

<img width="518" alt="Screen Shot 2022-10-05 at 3 23 47 PM" src="https://user-images.githubusercontent.com/21046714/194151832-f5054656-a147-498e-b90f-780caf300814.png">

## Notes

Universal Analytics keys will still work with `gtag.js` until July 2023. We currently have a UA key on OSHub staging but a GA4 key on production. 

## Testing Instructions

* Update your `.env` to set `REACT_APP_GOOGLE_ANALYTICS_KEY` equal to the new G-XXXXX (GA4 formatted) id. 
* In `public/index.html` replace `%REACT_APP_GOOGLE_ANALYTICS_KEY%` with the GA4 id 
     - This should happen automatically in production. See [docs](https://create-react-app.dev/docs/adding-custom-environment-variables/#referencing-environment-variables-in-the-html).
* Change `util.ga.js` line 35 to `if (!environment) {` (to allow analytics to run in development).
* Run `./scripts/server`
* Visit http://localhost:6543/settings and check if your analytics permissions are on or off. Turn them on (approve them) if they are off. 
* Open the network tab in the developer tools and refresh the page. 
* Search for requests containing the GA4 key. You should see several. In one of the payloads, you should see "User consented to GA tracking".
* Click around the site and you will see additional requests. 
* Go back to settings and reject analytics. Refresh the page. 
* Now in the network tab, you should only see two GET requests and no POSTs. Clicking around should send no POST requests with user data. 

## Checklist
